digitalstrom_mqtt/mqtt.go

```diff
@@ -194,6 +224,254 @@ func (dm *DigitalstromMqtt) getTopic(deviceType string, deviceId string, deviceN
 	return topic
 }
 
+// Returns MQTT topic to publish the Server status.
+func (dm *DigitalstromMqtt) getStatusTopic() string {
+	root_topic := strings.Split(dm.config.TopicFormat, "/")[0]
+	return root_topic + "/server/state"
+}
+
+// Publish the current binary status into the MQTT topic.
+func (dm *DigitalstromMqtt) publishDiscoveryMessages() {
+	for _, device := range dm.digitalstrom.GetAllDevices() {
+		messages, err := dm.deviceToHomeAssistantDiscoveryMessage(device)
+		if utils.CheckNoErrorAndPrint(err) {
+			for _, discovery_message := range messages {
+				dm.client.Publish(discovery_message.topic, 0, dm.config.Retain, discovery_message.message)
+			}
+		}
+	}
+	for _, circuit := range dm.digitalstrom.GetAllCircuits() {
+		messages, err := dm.circuitToHomeAssistantDiscoveryMessage(circuit)
+		if utils.CheckNoErrorAndPrint(err) {
+			for _, discovery_message := range messages {
+				dm.client.Publish(discovery_message.topic, 0, dm.config.Retain, discovery_message.message)
+			}
+		}
+	}
+}
+
+// Define a Home Assistant discovery message as its MQTT topic and the message
+// to be published.
+type HassDiscoveryMessage struct {
+	topic   string
+	message []byte
+}
+
+// Return the definition of the light and cover entities coming from a device.
+func (dm *DigitalstromMqtt) deviceToHomeAssistantDiscoveryMessage(device digitalstrom.Device) ([]HassDiscoveryMessage, error) {
+	// Check for device instances where the discovery message can not be created.
+	if device.Name == "" {
+		return nil, fmt.Errorf("empty device name, skipping discovery message")
+	}
+	if (device.DeviceType != digitalstrom.Light) && (device.DeviceType != digitalstrom.Blind) {
+		return nil, fmt.Errorf("device type not supported %s", device.DeviceType)
+	}
+	device_config := map[string]interface{}{
+		"configuration_url": "https://" + dm.ds_config.Host,
+		"identifiers":       []interface{}{device.Dsid, device.Dsuid},
+		"manufacturer":      "DigitalStrom",
+		"model":             device.HwInfo,
+		"name":              device.Name,
+	}
+	availability := []interface{}{
+		map[string]interface{}{
+			"topic":                 dm.getStatusTopic(),
+			"payload_available":     Online,
+			"payload_not_available": Offline,
+		},
+	}
+	message := map[string]interface{}{}
+	topic := ""
+	if device.DeviceType == digitalstrom.Light {
+		// Setup configuration for a MQTT Cover in Home Assistant:
+		// https://www.home-assistant.io/integrations/light.mqtt/
+		topic = dm.config.HomeAssistantDiscoveryPrefix + "/light/" + device.Dsid + "/light/config"
+		message = map[string]interface{}{
+			"device":            device_config,
+			"name":              utils.RemoveRegexp(device.Name, "light"),
```

Comment:
I don't like this `RemoveRegexp` thingy. This is just for your setup. For me it messes everything....

For my name I use `light_` and `blinds_`  not cover. See image below:

![Screenshot from 2022-03-18 09-07-57](https://user-images.githubusercontent.com/3659220/158961723-8787ef7b-2807-4882-9fc5-634d57408e38.png)

In my opinion, just keep the full name.

Reply:
True, I assumed names would be set with spaces and not dashes or underscores. My bad.

True you can always rename it on Home Assistant, but for normal setups you could end up with renaming 30 devices which can be tedious. What about making this optional through a config parameter, so you can put a regex there that will be remove from the name and therefore could also work with other types of names (like `blinds` in your case).

The idea here is that in DS you could have your device name be: `Living Room Light`, but using that as a name would create an entity in Home Assistant that is called: `light.living_room_light`. As `_light` is part of the component there is no need to have it duplicated.

Reply:
In my opinion, it's not a big deal to have `light.living_room_light` as name in home assistant.  Is it?

I'm ok with a config, but in my honest opinion, it's not worth it. Keep it simple and stupid like we said.

Reply:
Yes, but having large setups (which is likely using DigitalStrom) could be helpful to have the option to customise the naming a little.